src/main/java/it/gov/pagopa/reward/service/synchronous/op/CancelTrxSynchronousServiceImpl.java

```diff
@@ -38,58 +52,101 @@ public CancelTrxSynchronousServiceImpl(
+                    return initiative2OnboardingInfo;
+                });
+
+
+        return trxChecks
+                .flatMap(i2o -> findUserInitiativeCounter(initiativeId, i2o, trxDTO))
+                .switchIfEmpty(Mono.error(new InitiativeNotActiveException(String.format(RewardConstants.ExceptionMessage.INITIATIVE_NOT_ACTIVE_FOR_USER_MSG,initiativeId),syncTrxRequest2TransactionDtoMapper
+                        .apply(trxCancelRequest, initiativeId, List.of(RewardConstants.TRX_REJECTION_REASON_NO_INITIATIVE)))))
+                .flatMap(counters -> checkCounterOrEvaluateThenUpdate(trxDTO, initiativeId, new UserInitiativeCountersWrapper(counters.getEntityId(), new HashMap<>(Map.of(initiativeId, counters))), trxCancelRequest.getRewardCents()))
+                .map(rewardTransaction -> mapper.apply(trxDTO.getId(), initiativeId, rewardTransaction));
     }
 
-    // build using just new inputs, setting actual Reward to -1*rewardCents
-    public TransactionDTO buildRefundTrx(TransactionProcessed trx, String initiativeId) {
-        TransactionDTO refundTrx = new TransactionDTO();
 
-        refundTrx.setIdTrxAcquirer(trx.getIdTrxAcquirer());
-        refundTrx.setAcquirerCode(trx.getAcquirerCode());
-        refundTrx.setAcquirerId(trx.getAcquirerId());
-        refundTrx.setUserId(trx.getUserId());
-        refundTrx.setCorrelationId(trx.getCorrelationId());
-        refundTrx.setTrxChargeDate(trx.getTrxChargeDate().atZone(CommonConstants.ZONEID).toOffsetDateTime());
-        refundTrx.setRejectionReasons(trx.getRejectionReasons());
-        refundTrx.setChannel(trx.getChannel());
+    public void transformIntoRefundTrx(TransactionDTO trx, String initiativeId, String organizationId, long rewardCents) {
+        BigDecimal chargeRewardEur = CommonUtilities.centsToEuro(rewardCents);
+
+        TransactionProcessed chargeTrx = trx2processed(trx, initiativeId, organizationId, chargeRewardEur);
 
         // refund info
-        refundTrx.setId(trx.getId());
-        refundTrx.setOperationType(refundOperationType);
-        refundTrx.setOperationTypeTranscoded(OperationType.REFUND);
-        refundTrx.setTrxDate(OffsetDateTime.now());
-        refundTrx.setAmountCents(trx.getAmountCents());
-        refundTrx.setAmount(trx.getAmount());
-        refundTrx.setEffectiveAmount(BigDecimal.ZERO.setScale(2, RoundingMode.UNNECESSARY));
-
-        Map<String, RefundInfo.PreviousReward> previousRewardMap = null;
-        Reward reward = trx.getRewards().get(initiativeId);
-        if(reward != null) {
-            previousRewardMap = Map.of(initiativeId, new RefundInfo.PreviousReward(reward.getInitiativeId(), reward.getOrganizationId(), reward.getAccruedReward()));
+        trx.setOperationType(refundOperationType);
+        trx.setOperationTypeTranscoded(OperationType.REFUND);
+        trx.setTrxDate(OffsetDateTime.now());
+        trx.setEffectiveAmount(BigDecimal.ZERO.setScale(2, RoundingMode.UNNECESSARY));
```

Comment:
this could be a class level constant
`BigDecimal.ZERO.setScale(2, RoundingMode.UNNECESSARY)`